### PR TITLE
Pleroma-otp: init at 2.2.2

### PR DESCRIPTION
**Pull request by picnoir**
##### Motivation for this change

See #62820.

This PR provides:

- A `pleroma-otp` package, the official pleroma binary distribution.
- A pleroma NixOS module.
- A NixOS manual entry explaining how to setup this module.
- A NixOS VM test testing that module. (`nix-build -A nixosTests.pleroma` to run it).

Some random notes:

- Using a binary distribution is far from ideal, however, building this Elixir project turned out being quite complex. `pleroma-otp` is just a starting point, it'd be nice to compile the project from scratch ourselves in the future.
- The VM test required to patch the `toot` CLI. I'll try to upstream that patch. (Edit: https://github.com/ihabunek/toot/pull/180)
- My English spelling is generally pretty bad, please be as pedantic as you can while reviewing the pleroma manual entry.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/pleroma.nix [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        '';
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    users = {
+      users."${cfg.user}" = {
+        description = "Pleroma user";
+        home = "/var/lib/pleroma";
```
Comment (kampka) [inline nixos/modules/services/networking/pleroma.nix:66]: Please make this configurable.

File: nixos/tests/pleroma.nix [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+  hosts = nodes: ''
+    ${nodes.pleroma.config.networking.primaryIPAddress} pleroma.nixos.test
+    ${nodes.client.config.networking.primaryIPAddress} client.nixos.test
+  '';
+in import ./make-test-python.nix {
+  name = "pleroma";
+  nodes = {
+    client = { nodes, pkgs, config, ... }: {
+      security.pki.certificateFiles = [ "${tls-cert pkgs}/cert.pem" ];
+      console.keyMap = "fr-bepo";
```
Comment (andir) [inline nixos/tests/pleroma.nix:193]: Huh, that is uncommon and unexpected :)

File: nixos/tests/pleroma.nix [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+      security.pki.certificateFiles = [ "${tls-cert pkgs}/cert.pem" ];
+      networking.extraHosts = hosts nodes;
+      console.keyMap = "fr-bepo";
+      networking.firewall.enable = false;
+      environment.systemPackages = with pkgs; [
+        postgresql
+        (provision-db pkgs)
+        (provision-secrets pkgs)
+        (provision-user pkgs)
+        sudo
```
Comment (andir) [inline nixos/tests/pleroma.nix:212]: Sudo should not be here. It usually comes via the module system and is currently a default.

File: nixos/tests/pleroma.nix [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        strace
```
Comment (andir) [inline nixos/tests/pleroma.nix:213]: Why do we need strace?

File: nixos/tests/pleroma.nix [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+      url = "https://github.com/NinjaTrappeur/toot/commit/b4a4c30f41c0cb7e336714c2c4af9bc9bfa0c9f2.patch";
+      sha256 = "sha256-0xxNwjR/fStLjjUUhwzCCfrghRVts+fc+fvVJqVcaFg=";
+    }) ];
+  });
+
+  hosts = nodes: ''
+    ${nodes.pleroma.config.networking.primaryIPAddress} pleroma.nixos.test
+    ${nodes.client.config.networking.primaryIPAddress} client.nixos.test
+  '';
+in import ./make-test-python.nix {
```
Comment (andir) [inline nixos/tests/pleroma.nix:188]: If you changed this to being a function that expects `{ pkgs, ... }: ` you can move all the helper script definitions down here and get rid of passing pkgs for each of them.

File: nixos/modules/services/networking/pleroma.nix (hunk 1 above)
Comment (NinjaTrappeur) [reply nixos/modules/services/networking/pleroma.nix:66]: This PR is focussing on systemd, see `StateDirectory` in https://www.freedesktop.org/software/systemd/man/systemd.exec.html for more context.

I think pointing to the service state directory is a good default.

If you're worried about the default home value and really care about it, you can override that in your machine description conf using `mkForce`.

File: nixos/tests/pleroma.nix (hunk 2 above)
Comment (NinjaTrappeur) [reply nixos/tests/pleroma.nix]: French soft power :P 

(fixed).

File: nixos/tests/pleroma.nix (hunk 4 above)
Comment (NinjaTrappeur) [reply nixos/tests/pleroma.nix]: Not really, it was meant to be useful in case I'd need to interactively debug the test in the future.

Removed it.

File: nixos/tests/pleroma.nix (hunk 5 above)
Comment (NinjaTrappeur) [reply nixos/tests/pleroma.nix]: Good point.

Done.

**Comment by NinjaTrappeur** [comment]
Thanks for the feedback.

Addressed and forced push.

File: nixos/modules/services/networking/pleroma.nix (hunk 1 above)
Comment (kampka) [reply nixos/modules/services/networking/pleroma.nix:66]: I don't care about the home dir, I care about where my data is stored.
The fact that the "StateDirectory" is not meant to be configurable is another terrible idea if systemd fanboys forcing their unjustified opinions upon the world.
`mkForce` does nothing in that regard as it cannot change the root of `StateDirectory`

Comment (andir) [reply nixos/modules/services/networking/pleroma.nix:66]: The way I handle those things:

```
fileSystems."/var/lib/service" = {
   fsType = "none";
   options = "bind";
   device = "/somewhere/where/my/data/is/stored";
};
```

While I hate it not being configurable it also ensures that there is a defined set of directories that are being mutated by a daemon (and I do not have to guess or read more docs). If I do not care about the data I do not mount a proper volume there but if I do I usually store it elsewhere.

Comment (kampka) [reply nixos/modules/services/networking/pleroma.nix:66]: > While I hate it not being configurable it also ensures that there is a defined set of directories that are being mutated by a daemon (and I do not have to guess or read more docs).

Except in practice, it's not like that at all because you basically cannot know where data is stored if you can neither configure it nor want to look at source and docs. So in the end, you'll have to read the source code of at least the module to find out if it actually does what you expect it to do, or roll the dice on whether your backups, storage / disk monitoring is actually doing something or if you're just lucky enough never to notice.

Comment (andir) [reply nixos/modules/services/networking/pleroma.nix:66]: No, all services go to `/var/lib/<somewhere>` that is pretty easy to predict. Anything else is just wrong and not predictable. For backups you should be more explicit and there I partially agree with having the path in the module system. You should collect them from the services you care about or just backup all of `/var/lib`. It makes reasoning about a system a lot easier if things are following one rule - be it your or my favorite or not. I am sick of services that want to write to `/opt/somewhere` and others requiring a home dir in `/home` etc..

Comment (kampka) [reply nixos/modules/services/networking/pleroma.nix:66]: > You should collect them from the services you care about or just backup all of /var/lib

Only backing up all of `/var/lib` is currently an option, which will give you a lot of garbage in addition to the services you actually want, because every service throws their state files in `/var/lib`, even if `/var/cache` or `/var/run` is technically more correct.

Anyways, it's easier to re-implement a service on a flake than to change anything on NixOS and besides, there is no point in having a theoretical discussion about this on a pull request of a specific service.

Comment (NinjaTrappeur) [reply nixos/modules/services/networking/pleroma.nix:66]: > Only backing up all of /var/lib is currently an option, which will give you a lot of garbage in addition to the services you actually want, because every service throws their state files in /var/lib, even if /var/cache or /var/run is technically more correct.

Let's move back to this PR scope. Only uploads and static files will be stored in `/var/lib/pleroma`.

>  The fact that the "StateDirectory" is not meant to be configurable is another terrible idea if systemd fanboys forcing their unjustified opinions upon the world.

This is actually coming from Linux's FHS spec. You can learn more about it there: https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch05s08.html.

Feel free to use Andi's trick to store your data elsewhere.

This module won't store any state data in `/var/cache` nor in `/var/run`.

Comment (kampka) [reply nixos/modules/services/networking/pleroma.nix:66]: I am well aware of the FHS, that is why I know that `static files and uploads` belong to `/srv` and not `/var/lib`, as also stated in the footnote on the page you linked. 
Also quoting the FHS for NixOS is quite ironic in itself I'd argue.

Again, moot point as this will neither convince anyone not should it.

**Comment by NinjaTrappeur** [comment]
Just forced push a fix for some issues I noticed:

1. Add `aarch64` support. It's a bindist, we need to fetch a different distribution dependending of the target arch (for x compilation). Thanks to Flokli for the help.
2. Added a read-only option pointing to the state directory.
3. Added a guard making sure we're running on a aarch64 or x86-64 system for the tests.

File: nixos/modules/services/networking/pleroma.nix (hunk 1 above)
Comment (NinjaTrappeur) [reply nixos/modules/services/networking/pleroma.nix]: I just added a read only option pointing to the statedir, making easier to point to it in your machine description.

```
fileSystems."${config.services.pleroma.stateDir}" = {
   fsType = "none";
   options = "bind";
   device = "/somewhere/where/my/data/is/stored";
};

```

File: pkgs/servers/pleroma-otp/default.nix [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  installPhase = ''
+    mkdir $out
+    cp -r * $out'';
+
+  # Pleroma is using the project's root path (here the store path)
+  # as its TMPDIR.
+  # Patching it to move the tmp dir to the actual tmpdir
+  postFixup = ''
+    wrapProgram $out/bin/pleroma \
+      --set-default RELEASE_TMP "/tmp"
```
Comment (flokli) [inline pkgs/servers/pleroma-otp/default.nix:53]: This surely doesn't persist anything by accident that might be useful?

Asking for `PrivateTmp=yes` mostly ;-)

If so, we should open an upstream issue regardless, to just default to using tempdir mechanisms, and add it in the comment above.

File: nixos/tests/pleroma.nix [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+*/
+
+import ./make-test-python.nix ({ pkgs, ... }:
+  let
+  send-toot = pkgs.writeScriptBin "send-toot" ''
+    set -eux
+    # toot is using the requests library internally. This library
+    # sadly embed its own certificate store instead of relying on the
+    # system one. Overriding this pretty bad default behaviour.
+    export REQUESTS_CA_BUNDLE=/etc/ssl/certs/ca-certificates.crt
```
Comment (flokli) [inline nixos/tests/pleroma.nix:33]: Is this using its own bundled version of it, or the nixpkgs-provided version?
I wonder what happened to 768c970c8d339f76947976c31ad2c8723247a8f3, I'd assume requests was patched to use `
/etc/ssl/certs/ca-certificates.crt` instead…

**Comment by petabyteboy** [comment]
Hey there, I'm running a source-built version of Pleroma since a few months. A binary version is already a big improvement, but also feel free to use my source-built version if you can make it fit:
https://git.petabyte.dev/petabyteboy/nixfiles/src/branch/master/pkgs/pleroma

**Comment by NinjaTrappeur** [comment]
Waaat, that's great news!

Can I cherry pick your source build and default to it instead of this pleroma bindist half measure? 

I see v2.2.0 has just been released I'll bump that this weekend.

**Comment by petabyteboy** [comment]
I bumped my package to 2.2.0, but it's still running on Elixir 1.10, because it complained about `import_config` statements with Elixir 1.11 despite my config file not containing any (probably caused by https://github.com/elixir-lang/elixir/commit/b2270699e13556cd10164350109cf30796497356). 
It might be problematic to upstream my source-built package, because it relies on some unmaintained Elixir tooling that hasn't made its way to nixpkgs for a reason: For example it currently uses builtins.fetchGit, which will not work on Hydra. Someone would need to change mix2nix to prefetch the Git repositories, insert the hash into mix.nix and use fetchgit instead of builtins.fetchGit. 
I'd be happy to make changes and take care of updating the Pleroma package in the future, but I'm not able to maintain the new Elixir tooling that would be required to add source-built Pleroma.

File: nixos/tests/pleroma.nix (hunk 7 above)
Comment (NinjaTrappeur) [reply nixos/tests/pleroma.nix:33]: That's weird, it do not seem like `requests` gets overriden in the proccess: https://github.com/NixOS/nixpkgs/blob/54985aeaf714505611f83589e57236010baee408/pkgs/applications/misc/toot/default.nix#L17

However, if I remove this line, the test end up failing with:

```
client # Traceback (most recent call last):
client #   File "/nix/store/wsj5danxnka3zgiz155w1y02p4zax1n5-python3.8-urllib3-1.25.10/lib/python3.8/site-packages/urllib3/connectionpool.py", line 670, in urlopen
client #     httplib_response = self._make_request(
client #   File "/nix/store/wsj5danxnka3zgiz155w1y02p4zax1n5-python3.8-urllib3-1.25.10/lib/python3.8/site-packages/urllib3/connectionpool.py", line 381, in _make_request
client #     self._validate_conn(conn)
client #   File "/nix/store/wsj5danxnka3zgiz155w1y02p4zax1n5-python3.8-urllib3-1.25.10/lib/python3.8/site-packages/urllib3/connectionpool.py", line 978, in _validate_conn
client #     conn.connect()
client #   File "/nix/store/wsj5danxnka3zgiz155w1y02p4zax1n5-python3.8-urllib3-1.25.10/lib/python3.8/site-packages/urllib3/connection.py", line 362, in connect
client #     self.sock = ssl_wrap_socket(
client #   File "/nix/store/wsj5danxnka3zgiz155w1y02p4zax1n5-python3.8-urllib3-1.25.10/lib/python3.8/site-packages/urllib3/util/ssl_.py", line 384, in ssl_wrap_socket
client #     return context.wrap_socket(sock, server_hostname=server_hostname)
client #   File "/nix/store/gvkhgh813chlsvdsnxgwhv72xns08xcp-python3-3.8.6/lib/python3.8/ssl.py", line 500, in wrap_socket
client #     return self.sslsocket_class._create(
client #   File "/nix/store/gvkhgh813chlsvdsnxgwhv72xns08xcp-python3-3.8.6/lib/python3.8/ssl.py", line 1040, in _create
client #     self.do_handshake()
client #   File "/nix/store/gvkhgh813chlsvdsnxgwhv72xns08xcp-python3-3.8.6/lib/python3.8/ssl.py", line 1309, in do_handshake
client #     self._sslobj.do_handshake()
client # ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate (_ssl.c:1124)
client # 
client # During handling of the above exception, another exception occurred:
client # 
client # Traceback (most recent call last):
client #   File "/nix/store/nplv8hd6mljj3rymv7s6269q13y1j9gk-python3.8-requests-2.24.0/lib/python3.8/site-packages/requests/adapters.py", line 439, in send
client #     resp = conn.urlopen(
client #   File "/nix/store/wsj5danxnka3zgiz155w1y02p4zax1n5-python3.8-urllib3-1.25.10/lib/python3.8/site-packages/urllib3/connectionpool.py", line 726, in urlopen
client #     retries = retries.increment(
client #   File "/nix/store/wsj5danxnka3zgiz155w1y02p4zax1n5-python3.8-urllib3-1.25.10/lib/python3.8/site-packages/urllib3/util/retry.py", line 439, in increment
client #     raise MaxRetryError(_pool, url, error or ResponseError(cause))
client # urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='pleroma.nixos.test', port=443): Max retries exceeded with url: /api/v1/instance (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate (_ssl.c:1124)')))
client # 
client # During handling of the above exception, another exception occurred:
client # 
client # Traceback (most recent call last):
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/bin/.toot-wrapped", line 9, in <module>
client #     sys.exit(main())
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/lib/python3.8/site-packages/toot/console.py", line 550, in main
client #     run_command(app, user, command_name, args)
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/lib/python3.8/site-packages/toot/console.py", line 532, in run_command
client #     return fn(app, user, parsed_args)
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/lib/python3.8/site-packages/toot/commands.py", line 188, in login_cli
client #     app = create_app_interactive(instance=args.instance, scheme=args.scheme)
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/lib/python3.8/site-packages/toot/auth.py", line 43, in create_app_interactive
client #     return config.load_app(instance) or register_app(instance, scheme)
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/lib/python3.8/site-packages/toot/auth.py", line 15, in register_app
client #     instance = api.get_instance(domain, scheme)
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/lib/python3.8/site-packages/toot/api.py", line 281, in get_instance
client #     return http.anon_get(url).json()
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/lib/python3.8/site-packages/toot/http.py", line 62, in anon_get
client #     response = send_request(request)
client #   File "/nix/store/hbwji1fddg9ndpaxm95nzwcbx2m0rm16-toot-0.27.0/lib/python3.8/site-packages/toot/http.py", line 17, in send_request
client #     response = session.send(prepared, allow_redirects=allow_redirects, **settings)
client #   File "/nix/store/nplv8hd6mljj3rymv7s6269q13y1j9gk-python3.8-requests-2.24.0/lib/python3.8/site-packages/requests/sessions.py", line 643, in send
client #     r = adapter.send(request, **kwargs)
client #   File "/nix/store/nplv8hd6mljj3rymv7s6269q13y1j9gk-python3.8-requests-2.24.0/lib/python3.8/site-packages/requests/adapters.py", line 514, in send
client #     raise SSLError(e, request=request)
client # requests.exceptions.SSLError: HTTPSConnectionPool(host='pleroma.nixos.test', port=443): Max retries exceeded with url: /api/v1/instance (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate (_ssl.c:1124)')))
client: output: Creating config file at /root/.config/toot/config.json
Looking up instance info...

error: 
Traceback (most recent call last):
  File "/nix/store/nc5cfr15lavxwa8i334f1xnwsd08h019-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 899, in run_tests
    exec(tests, globals())
  File "<string>", line 1, in <module>
  File "<string>", line 7, in <module>
  File "/nix/store/nc5cfr15lavxwa8i334f1xnwsd08h019-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 422, in succeed
    raise Exception(
Exception: command `send-toot` failed (exit code 1)
cleaning up
killing client (pid 57)
killing pleroma (pid 9)
(0.00 seconds)
```

TL;DR: it's not reading my `/etc/ssl/certs` file.

I'll leave that line for now.

Comment (NinjaTrappeur) [reply nixos/tests/pleroma.nix:33]: Mystery solved: the patch is no more.

https://github.com/NixOS/nixpkgs/blob/54985aeaf714505611f83589e57236010baee408/pkgs/development/python-modules/requests/default.nix#L13

File: pkgs/servers/pleroma-otp/default.nix (hunk 6 above)
Comment (NinjaTrappeur) [reply pkgs/servers/pleroma-otp/default.nix:53]: Yup, it's safe.

I read the full 3 layers of OTP wrappers to double check :)

File: nixos/tests/pleroma.nix (hunk 7 above)
Comment (flokli) [reply nixos/tests/pleroma.nix:33]: Good point.

**Comment by NinjaTrappeur** [comment]
Bumped the PR to pleroma latest 2.2.0 release.

----------------------

Ack @petabyteboy . To be franc, I don't really have any personal use of the elixir/otp ecosystem outside of Pleroma. I'm not really willing to invest a lot of the little hobby time I have in this ecosystem for now.

Let's move on with the bindist for now and see if somebody is willing to improve the `mix.nix` situation.

I wish I discovered your repo sooner, it'd be very helpful. I just posted a comment on #62820 pointing to your source build to help with discoverability (in the secret hope to trigger more interest in upstreaming that :) ).

------------------------

Dear reader, feel free to review/test further this PR ;)

File: nixos/modules/services/networking/pleroma.nix [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        description = "Directory where the pleroma service will save the uploads and static files.";
+      };
+
+      configs = mkOption {
+        type = with types; listOf str;
+        description = ''
+          Pleroma public configuration.
+
+          This list gets appended from left to
+          write into /etc/pleroma/config.exs. Elixir evaluates its
```
Comment (JJJollyjim) [inline nixos/modules/services/networking/pleroma.nix:40]: `write` should be `right` here :)

**Review by JJJollyjim — CHANGES_REQUESTED** [review]

**Comment by JJJollyjim** [comment]
Looks good to me, would love to see it merged :)

File: nixos/modules/services/networking/pleroma.nix (hunk 8 above)
Comment (NinjaTrappeur) [reply nixos/modules/services/networking/pleroma.nix]: Oops, indeed :sweat_smile: 

Thanks for the attentive read :) Fixed.

File: nixos/tests/pleroma.nix [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+    pleroma_ctl user new jamy jamy@nixos.test --password 'jamy-password' --moderator --admin -y
+  '';
+
+  tls-cert = pkgs.runCommandNoCC "selfSignedCerts" { buildInputs = [ pkgs.openssl ]; } ''
+    openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -nodes -subj '/CN=pleroma.nixos.test'
+    mkdir -p $out
+    cp key.pem cert.pem $out
+  '';
+
+  # Toot is preventing users from feeding login_cli a password non
```
Comment (andir) [inline nixos/tests/pleroma.nix:171]: nit mind making the comment style consistent with the others?

File: nixos/tests/pleroma.nix [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+      services = {
+        pleroma = {
+          enable = true;
+          configs = [
+            pleroma-conf
+          ];
+        };
+        postgresql = {
+          enable = true;
+          package = pkgs.postgresql_12;
```
Comment (andir) [inline nixos/tests/pleroma.nix:219]: Was the default postgresql version not enough?

File: nixos/modules/services/networking/pleroma.nix [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+        type = types.str;
+        default = "/var/lib/pleroma";
+        readOnly = true;
+        description = "Directory where the pleroma service will save the uploads and static files.";
+      };
+
+      configs = mkOption {
+        type = with types; listOf str;
+        description = ''
+          Pleroma public configuration.
```
Comment (andir) [inline nixos/modules/services/networking/pleroma.nix:37]: Would it make sense to link to upstream documentation here? Note: I've never deployed any services in this language...

File: nixos/tests/pleroma.nix (hunk 10 above)
Comment (NinjaTrappeur) [reply nixos/tests/pleroma.nix:219]: There's no default, you have to be explicit here: https://github.com/NixOS/nixpkgs/blob/2df221ec8a95566ba771f102fd421c12b41d875c/nixos/modules/services/databases/postgresql.nix#L40

File: nixos/modules/services/networking/pleroma.nix (hunk 11 above)
Comment (NinjaTrappeur) [reply nixos/modules/services/networking/pleroma.nix:37]: :+1:

I added a link to the upstream documentation in the manual pleroma entry. I also added a small note to this NixOS option inviting the user to read the NixOS manual relevant section.

File: nixos/tests/pleroma.nix (hunk 9 above)
Comment (NinjaTrappeur) [reply nixos/tests/pleroma.nix]: Done

**Review by andir — APPROVED** [review]

**Comment by NinjaTrappeur** [comment]
I'm not sure what happened on ofborg here. The x86 test & aarch64 tests ran successfully but the x86 `pleroma-otp` build seemed to have failed. It makes no sense to me: I'm running this PR on my Pleroma instance running on a x86 machine.

--------------------------

I've been running this PR for the last 3 months on my personal Pleroma instance without any issue. This PR also received an approval. 

I'm going to merge merge this in a week from now. If you have any objections, do holler now.

**Comment by b123400** [comment]
Hello, I just tried to build with the files and got an interesting error

```
building '/nix/store/mq7iw7yjpqqbs6pxihcqmx5jdzdc7gm7-download.drv'...

trying https://git.pleroma.social/pleroma/pleroma/-/jobs/175284/artifacts/download
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 55.9M  100 55.9M    0     0   9.7M      0  0:00:05  0:00:05 --:--:-- 12.4M
hash mismatch in fixed-output derivation '/nix/store/wivbhgbr5l9izc9nm13dch4b2qg7wk75-download':
  wanted: sha256:0z93qvyda5kh794svk00pgbzhm5pv663rjkzkwwb6sfk7f9287ir
  got:    sha256:1c6l04gga9iigm249ywwcrjg6wzy8iiid652mws3j9dnl71w2sim
```

nix-prefetch-url gives the same result:

```
# nix-prefetch-url https://git.pleroma.social/pleroma/pleroma/-/jobs/175284/artifacts/download
[55.9 MiB DL]
path is '/nix/store/wivbhgbr5l9izc9nm13dch4b2qg7wk75-download'
1c6l04gga9iigm249ywwcrjg6wzy8iiid652mws3j9dnl71w2sim
```

Can you confirm?

**Comment by NinjaTrappeur** [comment]
Yup, I can confirm, thanks for the heads up!

I forgot or failed the copy paste while updating to 2.2.2 on x86_64. The FOD did not fail on my computer since I had 2.2.1 with the correct hash in my nix-store. Maybe Ofborg was failing because of hash mismatch in the end.

**Comment by b123400** [comment]
Thank you! Looking forward to the merge :)

**Comment by happysalada** [comment]
You might be interested to know that an initial effort has been started to package mix project.
https://github.com/NixOS/nixpkgs/pull/112477
It's not quite a mix nix, as it is a little more simplistic, but it should be available faster.
I wish I had known about this PR earlier, there are lots of interesting things, great work!
